File: src/HolzShots/Input/Actions/EditImageCommand.vb
Imports HolzShots.Composition.Command
Imports HolzShots.UI

Namespace Input.Actions
    <Command("editImage")>
    Public Class EditImageCommand
        Inherits FileDependentCommand
        Implements ICommand(Of HSSettings)

        Private Const OpenInShotEditor = "Open Image in ShotEditor"

        Public Function Invoke(parameters As IReadOnlyDictionary(Of String, String), settingsContext As HSSettings) As Task Implements ICommand(Of HSSettings).Invoke
            If parameters Is Nothing Then Throw New ArgumentNullException(NameOf(parameters))
            If settingsContext Is Nothing Then Throw New ArgumentNullException(NameOf(settingsContext))

            Dim fileName = If(
                parameters.Count <> 1 OrElse Not parameters.ContainsKey(FileNameParameter),
                ShowFileSelector(OpenInShotEditor),
                parameters(FileNameParameter)
            )

            If fileName Is Nothing Then Return Task.CompletedTask ' We did not get a valid file name (user cancelled or something else was strange)

            If Not CanProcessFile(fileName) Then
                ' TODO: Error Message
                Return Task.CompletedTask
            End If

            Dim bmp As New Bitmap(fileName)
            Dim shot = Screenshot.FromImported(bmp)
            Dim editor As New ShotEditor(shot, settingsContext)
            AddHandler editor.Disposed, Sub() bmp.Dispose()
            editor.Show()

            Return Task.CompletedTask
        End Function
    End Class
End Namespace

File: src/HolzShots.New/Input/Action/EditImageCommand.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using HolzShots.Composition.Command;
using System.Drawing;

namespace HolzShots.Input.Actions
{
    [Command("editImage")]
    public class EditImageCommand : FileDependentCommand, ICommand<HSSettings>
    {
        private const string OpenInShotEditor = "Open Image in ShotEditor";

        public Task Invoke(IReadOnlyDictionary<string, string> parameters, HSSettings settingsContext)
        {
            if (parameters == null)
                throw new ArgumentNullException(nameof(parameters));
            if (settingsContext == null)
                throw new ArgumentNullException(nameof(settingsContext));

            var fileName = parameters.Count != 1 || !parameters.ContainsKey(FileNameParameter) ? ShowFileSelector(OpenInShotEditor) : parameters[FileNameParameter];

            if (fileName == null)
                return Task.CompletedTask; // We did not get a valid file name (user cancelled or something else was strange)

            if (!CanProcessFile(fileName))
                return Task.CompletedTask; // TODO: Error Message

            var bmp = new Bitmap(fileName);
            var shot = Screenshot.FromImported(bmp);
            var editor = new UI.ShotEditor(shot, New.HolzShotsApplication.Instance.Uploaders, settingsContext);
            editor.Disposed += (s, e) => bmp.Dispose();
            editor.Show();

            return Task.CompletedTask;
        }
    }
}
